src/renderer.nim

```diff
@@ -0,0 +1,173 @@
+import geometry, pcg, hdrimage, camera, shapes, scene, hitrecord
+
+from std/strformat import fmt
+from std/math import cos, sin, sqrt, PI
+import std/[options, strutils, terminal]
+from std/math import cos, sin, sqrt, PI
+
+
+type
+    RendererKind* = enum
+        rkOnOff, rkFlat, rkPathTracer
+
+    Renderer* = object
+        camera*: Camera
+
+        case kind*: RendererKind
+        of rkFlat: discard
+
+        of rkOnOff:
+            hitCol*: Color
+
+        of rkPathTracer:
+            nRays*, maxDepth*, rouletteLim*: int
+
+
+proc newOnOffRenderer*(camera: Camera, hitCol = WHITE): Renderer {.inline.} =
+    Renderer(kind: rkOnOff, camera: camera, hitCol: hitCol,)
+
+proc newFlatRenderer*(camera: Camera): Renderer {.inline.} =
+    Renderer(kind: rkFlat, camera: camera)
+
+proc newPathTracer*(camera: Camera, nRays = 25, maxDepth = 10, rouletteLim = 3): Renderer {.inline.} =
+    Renderer(kind: rkPathTracer, camera: camera, nRays: nRays, maxDepth: maxDepth, rouletteLim: rouletteLim)
+
+
+proc displayProgress(current, total: int) =
+    let
+        percentage = int(100 * current / total)
+        progress = 50 * current div total
+        bar = "[" & "#".repeat(progress) & "-".repeat(50 - progress) & "]"
+        color = if percentage <= 50: fgYellow else: fgGreen
+
+    stdout.eraseLine
+    stdout.styledWrite(fgWhite, "Rendering progress: ", fgRed, "0% ", fgWhite, bar, color, fmt" {percentage}%")
+    stdout.flushFile
+
+
+proc scatterRay*(refSystem: ReferenceSystem, brdf: BRDF, ray: Ray, rg: var PCG): Ray =
+    case brdf.kind:
+    of DiffuseBRDF:
+        let 
+            cos2 = rg.rand
+            (c, s) = (sqrt(cos2), sqrt(1 - cos2))
+            phi = 2 * PI * rg.rand
+        
+        Ray(
+            origin: ORIGIN3D, 
+            dir: refSystem.fromCoeff([float32 c * cos(phi), c * sin(phi), s]), 
+            tSpan: (float32 1e-3, float32 Inf), 
+            depth: ray.depth + 1
+        )
+
+    of SpecularBRDF: 
+        Ray(
+            origin: ORIGIN3D,
+            dir: ray.dir.normalize - 2 * dot(refSystem.base[2], ray.dir.normalize) * refSystem.base[2],
+            tspan: (float32 1e-3, float32 Inf), 
+            depth: ray.depth + 1
+        )
+
+
+proc sampleRay(renderer: Renderer; scene: Scene, subScene: SubScene, ray: Ray, maxShapesPerLeaf: int, rg: var PCG): Color =
+    result = scene.bgCol
+
+    let hitLeafNodes = subScene.getHitLeafs(ray)
+    if hitLeafNodes.isSome:
+
+        case renderer.kind
+        of rkOnOff:
+            if newHitRecord(subScene.rs, hitLeafNodes.get, ray).isSome:
```

Comment:
Maybe newHitRecord could be located outside of caseOff, but sometimes in ```ofPathTracer``` we don't even have to compute it so it wouldn't be a wise computational choice